src/msw/overlay.cpp

```diff
@@ -100,8 +99,16 @@ void wxOverlayImpl::Init(wxDC* dc, int , int , int , int )
 
     m_window = dc->GetWindow();
 
-    m_rect.SetSize(m_window->GetClientSize());
-    m_rect.SetPosition(m_window->GetScreenPosition());
+    if ( wxDynamicCast(dc, wxClientDC) )
+    {
+        m_rect = m_window->GetClientRect();
+        // The rectangle should be in screen coordinates
+        m_window->ClientToScreen(&m_rect.x, &m_rect.y);
+    }
+    else // wxWindowDC
+    {
+        m_rect = m_window->GetScreenRect();
+    }
```

Comment:
Why can't we always just use `GetScreenRect()` here? Is it just because we really need client size for `wxClientDC` but full window size otherwise? If so, we could still use something like

```c++
m_rect = wxRect(m_window->GetScreenPosition(), dc->GetSize());
```

I think, without any dynamic casts.

Reply:
When `wxClientDC` is used we want the overlay window to be positioned over the client area of the window. `m_window->GetScreenPosition()` returns the upper left corner of the window and not the top left position of the client area.

The overlay window positioned at `m_window->GetScreenPosition()` (**incorrect**): 

![overlay_incorrect](https://github.com/wxWidgets/wxWidgets/assets/7704771/d5b04847-3eab-4bb1-ad52-1c0c8080d29c)

The overlay window positioned at `m_window->GetClientRect()` (**correct**):

![overlay_correct](https://github.com/wxWidgets/wxWidgets/assets/7704771/18e80882-2fc2-4e26-b797-6a899be00759)

Reply:
Ah, I see, thanks for the explanation!

I'd still like to avoid dynamic cast, for all the usual reasons. As always, we can just add a virtual function instead and while it's a bit ugly to have window-oriented function in `wxDC`, we do already have `GetWindow()` there, so we could also add some `GetOriginInWindow()` or `GetWindowOffset()` or whatever else you'd prefer that would return `wxPoint(0, 0)` by default but would be overridden to return `GetClientAreaOrigin()` in `wxClientDCImpl`.

Do you see any problems with doing it like this? Thanks again!